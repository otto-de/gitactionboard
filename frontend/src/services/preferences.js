import storageService from "@/services/storageService";

class Preferences {
  __get__(configKey) {
    return storageService.getItem(configKey);
  }

  __set__(configKey, value) {
    storageService.setItem(configKey, value);
  }

  get showHealthyBuilds() {
    const showHealthBuildConfig = this.__get__("show-healthy-builds");
    return !(showHealthBuildConfig && showHealthBuildConfig === "false");
  }

  set showHealthyBuilds(value) {
    this.__set__("show-healthy-builds", value);
  }

  get disableIdleOptimization() {
    return this.__get__("disable-max-idle-time") === "true";
  }

  set disableIdleOptimization(value) {
    this.__set__("disable-max-idle-time", value);
  }

  get maxIdleTime() {
    const maxIdleTimeConfig = this.__get__("max-idle-time");
    return maxIdleTimeConfig ? parseInt(maxIdleTimeConfig) : 5;
  }

  set maxIdleTime(value) {
    this.__set__("max-idle-time", value);
  }

  get hiddenElements() {
    return JSON.parse(this.__get__("hidden-elements") || "{}");
  }

  set hiddenElements(value) {
    this.__set__("hidden-elements", JSON.stringify(value));
  }
}

export default new Preferences();
