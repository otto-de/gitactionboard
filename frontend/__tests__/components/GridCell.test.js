import { describe, expect, it, beforeEach, vi } from 'vitest';
import { mount, requestAnimationFrameAsPromise, retryUntil } from '../test-utils';
import GridCell from '@/components/GridCell.vue';
import preferences from '@/services/preferences';

describe('<GridCell />', () => {
  const defaultProps = {
    name: 'webpack-cli :: webpack-cli :: Lint Commit Messages',
    url: 'https://github.com/webpack/webpack-cli/runs/7858502725',
    lastExecutedTime: '2022-08-16T03:45:02.000Z',
    buildMonitorViewEnabled: true
  };

  const rootId = `${defaultProps.name.replaceAll(/[\\:\s]/g, '-')}`;

  beforeEach(() => {
    vi.spyOn(preferences, 'theme', 'get').mockReturnValueOnce('light');
  });

  it.each([
    [undefined, undefined, undefined, undefined, undefined],
    [false, undefined, undefined, undefined, undefined],
    [true, undefined, undefined, undefined, undefined],
    [false, 'success', undefined, undefined, undefined],
    [false, 'SUCCESS', undefined, undefined, undefined],
    [false, 'unknown', undefined, undefined, undefined],
    [false, 'FAILURE', undefined, undefined, undefined],
    [false, 'random', undefined, undefined, undefined],
    [true, 'success', undefined, undefined, undefined],
    [true, 'SUCCESS', undefined, undefined, undefined],
    [true, 'unknown', undefined, undefined, undefined],
    [true, 'FAILURE', undefined, undefined, undefined],
    [true, 'random', undefined, undefined, undefined],
    [false, 'failure', undefined, undefined, true],
    [false, 'success', undefined, undefined, true],
    [true, 'failure', undefined, undefined, true],
    [true, 'success', undefined, undefined, true],
    [true, 'failure', undefined, undefined, false],
    [true, 'success', undefined, undefined, false],
    [false, 'failure', false, true, true]
  ])(
    'should render grid cell when inProgress=%s, status=%s, hidden=%s, displayToggleVisibility=%s, showRelativeTime=%s',
    (inProgress, status, hidden, displayToggleVisibility, showRelativeTime) => {
      const wrapper = mount(GridCell, {
        props: { ...defaultProps, inProgress, status, hidden, displayToggleVisibility, showRelativeTime }
      });
      expect(wrapper.html()).toMatchSnapshot();
    });

  it('should render grid cell when build monitor view is disabled', () => {
    const wrapper = mount(GridCell, {
      props: { ...defaultProps, buildMonitorViewEnabled: false }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render in progress grid cell when build monitor view is disabled', () => {
    const wrapper = mount(GridCell, {
      props: { ...defaultProps, inProgress: true, buildMonitorViewEnabled: false }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it.each([
    [true, true, true],
    [false, true, true],
    [true, false, true],
    [false, false, true],
    [true, true, false],
    [false, false, false],
    [false, true, false], [true, false, false]
  ])('should flip on hover when hidden=%s, displayToggleVisibility=%s, showRelativeTime=%s',
    async (hidden, displayToggleVisibility, showRelativeTime) => {
      const wrapper = mount(GridCell, {
        props: { ...defaultProps, hidden, displayToggleVisibility, showRelativeTime }
      });
      await wrapper.find(`#${defaultProps.name.replaceAll(/[\\:\s]/g, '-')}`).trigger('mouseenter');
      await retryUntil(async () => {
        await requestAnimationFrameAsPromise();
        expect(wrapper.find(`[test-id="${rootId}-toolbar"]`).exists()).toBeTruthy();
      });
      expect(wrapper.html()).toMatchSnapshot();
    });

  it.each([
    [true, true, true],
    [false, true, true],
    [true, false, true],
    [false, false, true],
    [true, true, false],
    [false, false, false],
    [false, true, false], [true, false, false]
  ])('should flip back once mouse is moved away when hidden=%s, displayToggleVisibility=%s, showRelativeTime=%s',
    async (hidden, displayToggleVisibility, showRelativeTime) => {
      const wrapper = mount(GridCell, {
        props: { ...defaultProps, hidden, displayToggleVisibility, showRelativeTime }
      });

      await wrapper.find(`#${rootId}`).trigger('mouseenter');
      await retryUntil(async () => {
        await requestAnimationFrameAsPromise();
        expect(wrapper.find(`[test-id="${rootId}-toolbar"]`).exists()).toBeTruthy();
      });

      await wrapper.find(`#${rootId}`).trigger('mouseleave');
      await retryUntil(async () => {
        await requestAnimationFrameAsPromise();
        expect(wrapper.find(`[test-id="${rootId}-toolbar"]`).exists()).toBeFalsy();
      });
      expect(wrapper.html()).toMatchSnapshot();
    });

  it('should emit toggle visibility event when clicked on toggle icon', async () => {
    const wrapper = mount(GridCell, {
      props: { ...defaultProps, displayToggleVisibility: true }
    });

    await wrapper.find(`#${defaultProps.name.replaceAll(/[\\:\s]/g, '-')}`).trigger('mouseenter');
    await retryUntil(async () => {
      await requestAnimationFrameAsPromise();
      expect(wrapper.find(`[test-id="${rootId}-change-visibility-icon"]`).exists()).toBeTruthy();
    });

    await wrapper.find(`[test-id="${rootId}-change-visibility-icon"]`).trigger('click');
    const emitted = wrapper.emitted();
    expect(emitted).toHaveProperty('toggleVisibility');
    expect(emitted.toggleVisibility).toEqual([[defaultProps.name]]);
  });
});
