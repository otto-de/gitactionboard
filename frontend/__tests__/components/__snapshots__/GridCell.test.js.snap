// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<GridCell /> > Toggle Visibility > should display toggle visibility icon with hide icon when hidden passed as false 1`] = `
<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated grid-cell" id="webpack-cli----webpack-cli----Lint-Commit-Messages" style="height: 90px;">
  <!---->
  <div class="v-card__loader">
    <div aria-hidden="true" aria-valuemax="100" aria-valuemin="0" class="v-progress-linear v-theme--light v-locale--is-ltr" role="progressbar" style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);">
      <!---->
      <div class="v-progress-linear__background" style="width: 100%;"></div>
      <transition-stub appear="false" css="true" name="fade-transition" persisted="false">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-text grid-cell-name pa-0 ma-0">
    <a aria-describedby="v-tooltip-0" href="https://github.com/webpack/webpack-cli/runs/7858502725" id="webpack-cli----webpack-cli----Lint-Commit-Messages-url" target="_blank" targetref="[object Object]">webpack-cli :: webpack-cli :: Lint Commit Messages</a>
    <!--teleport start-->
    <!--teleport end-->
  </div>
  <div class="d-flex">
    <!--v-if-->
    <div class="v-spacer"></div>
    <i aria-describedby="v-tooltip-1" aria-hidden="false" class="mdi-eye-off mdi v-icon notranslate v-theme--light v-icon--size-small v-icon--clickable" role="button" targetref="[object Object]" test-id="webpack-cli----webpack-cli----Lint-Commit-Messages-change-visibility-icon"></i>
    <!--teleport start-->
    <!--teleport end-->
  </div>
  <!---->
  <!---->
  <span class="v-card__underlay"></span>
</div>
`;

exports[`<GridCell /> > Toggle Visibility > should display toggle visibility icon with hide icon when hidden passed as true 1`] = `
<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated grid-cell" id="webpack-cli----webpack-cli----Lint-Commit-Messages" style="height: 90px;">
  <!---->
  <div class="v-card__loader">
    <div aria-hidden="true" aria-valuemax="100" aria-valuemin="0" class="v-progress-linear v-theme--light v-locale--is-ltr" role="progressbar" style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);">
      <!---->
      <div class="v-progress-linear__background" style="width: 100%;"></div>
      <transition-stub appear="false" css="true" name="fade-transition" persisted="false">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-text grid-cell-name pa-0 ma-0">
    <a aria-describedby="v-tooltip-0" href="https://github.com/webpack/webpack-cli/runs/7858502725" id="webpack-cli----webpack-cli----Lint-Commit-Messages-url" target="_blank" targetref="[object Object]">webpack-cli :: webpack-cli :: Lint Commit Messages</a>
    <!--teleport start-->
    <!--teleport end-->
  </div>
  <div class="d-flex">
    <!--v-if-->
    <div class="v-spacer"></div>
    <i aria-describedby="v-tooltip-1" aria-hidden="false" class="mdi-eye mdi v-icon notranslate v-theme--light v-icon--size-small v-icon--clickable" role="button" targetref="[object Object]" test-id="webpack-cli----webpack-cli----Lint-Commit-Messages-change-visibility-icon"></i>
    <!--teleport start-->
    <!--teleport end-->
  </div>
  <!---->
  <!---->
  <span class="v-card__underlay"></span>
</div>
`;

exports[`<GridCell /> > Toggle Visibility > should display toggle visibility icon with hide icon when hidden passed as undefined 1`] = `
<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated grid-cell" id="webpack-cli----webpack-cli----Lint-Commit-Messages" style="height: 90px;">
  <!---->
  <div class="v-card__loader">
    <div aria-hidden="true" aria-valuemax="100" aria-valuemin="0" class="v-progress-linear v-theme--light v-locale--is-ltr" role="progressbar" style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);">
      <!---->
      <div class="v-progress-linear__background" style="width: 100%;"></div>
      <transition-stub appear="false" css="true" name="fade-transition" persisted="false">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-text grid-cell-name pa-0 ma-0">
    <a aria-describedby="v-tooltip-0" href="https://github.com/webpack/webpack-cli/runs/7858502725" id="webpack-cli----webpack-cli----Lint-Commit-Messages-url" target="_blank" targetref="[object Object]">webpack-cli :: webpack-cli :: Lint Commit Messages</a>
    <!--teleport start-->
    <!--teleport end-->
  </div>
  <div class="d-flex">
    <!--v-if-->
    <div class="v-spacer"></div>
    <i aria-describedby="v-tooltip-1" aria-hidden="false" class="mdi-eye-off mdi v-icon notranslate v-theme--light v-icon--size-small v-icon--clickable" role="button" targetref="[object Object]" test-id="webpack-cli----webpack-cli----Lint-Commit-Messages-change-visibility-icon"></i>
    <!--teleport start-->
    <!--teleport end-->
  </div>
  <!---->
  <!---->
  <span class="v-card__underlay"></span>
</div>
`;

exports[`<GridCell /> > should render grid cell 1`] = `
<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated grid-cell" id="webpack-cli----webpack-cli----Lint-Commit-Messages" style="height: 90px;">
  <!---->
  <div class="v-card__loader">
    <div aria-hidden="true" aria-valuemax="100" aria-valuemin="0" class="v-progress-linear v-theme--light v-locale--is-ltr" role="progressbar" style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);">
      <!---->
      <div class="v-progress-linear__background" style="width: 100%;"></div>
      <transition-stub appear="false" css="true" name="fade-transition" persisted="false">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-text grid-cell-name pa-0 ma-0">
    <a aria-describedby="v-tooltip-0" href="https://github.com/webpack/webpack-cli/runs/7858502725" id="webpack-cli----webpack-cli----Lint-Commit-Messages-url" target="_blank" targetref="[object Object]">webpack-cli :: webpack-cli :: Lint Commit Messages</a>
    <!--teleport start-->
    <!--teleport end-->
  </div>
  <div class="d-flex">
    <!--v-if-->
    <div class="v-spacer"></div>
    <!--v-if-->
  </div>
  <!---->
  <!---->
  <span class="v-card__underlay"></span>
</div>
`;
